File: crates/compiler/frontend/typescript/src/program-instance.ts
// Defines typed generic program composition operands.

export type ProgramRef = {
  program_ref: string;
};

export type ProgramInstanceRef = {
  instance_node_id: string;
  program_ref: string;
};

export type ProgramNewSpec = {
  program_ref: string;
  initial_context?: Record<string, unknown>;
};

export type ProgramInvokeSpec = {
  receiver: ProgramRef | ProgramInstanceRef;
  input?: Record<string, unknown>;
};

/** Convert an instance reference to its canonical receiver operand. */
export function programInstanceReceiver(
  instance: ProgramInstanceRef,
): Record<string, string> {
  return { program_instance_ref: instance.instance_node_id };
}

/** Convert a typed creation request to canonical operands. */
export function programNewOperands(spec: ProgramNewSpec): Record<string, unknown> {
  const operands: Record<string, unknown> = { program_ref: spec.program_ref };
  if (spec.initial_context !== undefined) operands.initial_context = spec.initial_context;
  return operands;
}

/** Convert a typed invocation request to canonical operands. */
export function programInvokeOperands(spec: ProgramInvokeSpec): Record<string, unknown> {
  const receiver =
    "instance_node_id" in spec.receiver
      ? programInstanceReceiver(spec.receiver)
      : { program_ref: spec.receiver.program_ref };
  const operands: Record<string, unknown> = { receiver };
  if (spec.input !== undefined) operands.input = spec.input;
  return operands;
}

File: crates/compiler/frontend/python/apxm_program/program_instance.py
"""Typed program composition handles for program.new and program.invoke."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Optional


@dataclass(frozen=True, slots=True)
class ProgramRef:
    """One-shot receiver for program.invoke against an imported program."""

    program_ref: str


@dataclass(frozen=True, slots=True)
class ProgramInstanceRef:
    """Opaque typed handle produced by program.new for stateful program.invoke."""

    instance_node_id: str
    program_ref: str

    def receiver_operand(self) -> dict[str, str]:
        return {"program_instance_ref": self.instance_node_id}


@dataclass(frozen=True, slots=True)
class ProgramNewSpec:
    """Typed operands recorded on a program.new semantic operation."""

    program_ref: str
    initial_context: Optional[dict[str, Any]] = None

    def to_operands(self) -> dict[str, Any]:
        operands: dict[str, Any] = {"program_ref": self.program_ref}
        if self.initial_context is not None:
            operands["initial_context"] = self.initial_context
        return operands


@dataclass(frozen=True, slots=True)
class ProgramInvokeSpec:
    """Typed operands recorded on a program.invoke semantic operation."""

    receiver: ProgramRef | ProgramInstanceRef
    input_payload: Optional[dict[str, Any]] = None

    def to_operands(self) -> dict[str, Any]:
        if isinstance(self.receiver, ProgramInstanceRef):
            receiver = self.receiver.receiver_operand()
        else:
            receiver = {"program_ref": self.receiver.program_ref}
        operands: dict[str, Any] = {"receiver": receiver}
        if self.input_payload is not None:
            operands["input"] = self.input_payload
        return operands
